framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1822,19 +1823,95 @@ protected void commonAssignmentCheck(
+     *     types.
+     */
+    private boolean shouldPrintVerbose(AnnotatedTypeMirror atm1, AnnotatedTypeMirror atm2) {
+        String atm1ToString = atm1.toString();
+        String atm2ToString = atm2.toString();
+        // If both types as strings are the same, use verbose toString.
+        if (atm2ToString.equals(atm1ToString)
+                // or if neither string contains an annotation
+                || (!atm2ToString.contains("@") && !atm1ToString.contains("@"))) {
+            return true;
+        }
+
+        SimpleAnnotatedTypeScanner<Boolean, Void> checkForMismatchedToStrings =
+                new SimpleAnnotatedTypeScanner<Boolean, Void>() {
+                    Map<String, List<AnnotatedTypeMirror>> map = new HashMap<>();
+
+                    @Override
+                    protected Boolean reduce(Boolean r1, Boolean r2) {
+                        if (r1 == null) {
+                            if (r2 == null) {
+                                return false;
+                            } else {
+                                return r2;
+                            }
+                        } else if (r2 == null) {
+                            return r1;
+                        } else {
+                            return r1 || r2;
+                        }
+                    }
+
+                    @Override
+                    protected Boolean defaultAction(AnnotatedTypeMirror type, Void avoid) {
+                        if (type == null) {
+                            return false;
+                        }
+                        String simple = type.toString();
+                        List<AnnotatedTypeMirror> list = map.get(simple);
+                        if (list == null) {
+                            list = new ArrayList<>();
+                            map.put(simple, list);
+                        }
+                        String verbose = type.toString(true);
+                        for (AnnotatedTypeMirror other : list) {
+                            String otherVerbose = other.toString(true);
+                            if (!verbose.equals(otherVerbose)) {
+                                list.add(type);
+                                return true;
+                            }
+                        }
+                        list.add(type);
+                        return false;
+                    }
+                };
+        Boolean result1 = checkForMismatchedToStrings.visit(atm1);
+        // SimpleAnnotatedTypeScanner#scan returns null is it encounters a null AnnotatedTypeMirror
+        if (result1 != null && result1) {
+            return true;
+        }
+        // Call reset to clear the visitor history, but not the map for Strings to types.
+        checkForMismatchedToStrings.reset();
+        Boolean result2 = checkForMismatchedToStrings.visit(atm2);
+        return result2 == null ? false : result2;
```

Comment:
Can you add a comment, somewhere, indicating under what condition visit returns null?